### ISSUE TITLE
Quantization_2of4_sparse_w4a16 Shows Significantly Degraded Performance

### ISSUE THREAD

**Issue by wangwenmingaa** [vllm-project/llm-compressor#1638]
**Describe the bug**
Referring to the code in the quantization_2of4_sparse_w4a16 example (https://github.com/vllm-project/llm-compressor/tree/main/examples/quantization_2of4_sparse_w4a16), the following steps were performed:
1. Sparsification: The model was first sparsified. The resulting sparsified model was saved in the sparsity_stage folder.

2. Fine-tuning: The model from the sparsity_stage folder was then fine-tuned. The fine-tuning code is as provided:    `
      from pathlib import Path
      import torch
      from loguru import logger
      from transformers import AutoModelForCausalLM, AutoTokenizer
      from llmcompressor import oneshot, train
      import os
      dataset = "ultrachat-200k"
      # Select the recipe for 2 of 4 sparsity and 4-bit activation quantization
      recipe = "2of4_w4a16_group-128_recipe.yaml"
      
      # save location of quantized model
      output_dir = "/mnt4/ckpt/Qwen3-32B-2of4_w4a16_0707_1"
      output_path = Path(output_dir)
      
      # set dataset config parameters
      splits = {"calibration": "train_gen[:10%]", "train": "train_gen"}
      max_seq_length = 512
      num_calibration_samples = 512
    
    # set training parameters for finetuning
    num_train_epochs = 1
    logging_steps = 500
    save_steps = 2000
    gradient_checkpointing = True  # saves memory during training
    learning_rate = 0.0001
    bf16 = True  # using full precision for training
    lr_scheduler_type = "cosine"
    warmup_ratio = 0.1
    preprocessing_num_workers = 32
    oneshot_kwargs = dict(
        dataset=dataset,
        recipe=recipe,
        num_calibration_samples=num_calibration_samples,
        preprocessing_num_workers=preprocessing_num_workers,
        splits=splits
    )
    training_kwargs = dict(
        bf16=bf16,
        optim="adamw_bnb_8bit",
        max_seq_length=max_seq_length,
        num_train_epochs=num_train_epochs,
        logging_steps=logging_steps,
        save_steps=save_steps,
        gradient_checkpointing=gradient_checkpointing,
        learning_rate=learning_rate,
        lr_scheduler_type=lr_scheduler_type,
        per_device_train_batch_size=4,
        warmup_ratio=warmup_ratio,
        torch_empty_cache_steps=100,
    )
    
    # Sparse finetune
    train(
        model=(output_path / "sparsity_stage"),
        **oneshot_kwargs,
        **training_kwargs,
        output_dir=output_dir,
        stage="finetuning_stage",
    )`

3. Checkpoint Selection: Due to the long duration of fine-tuning, the model from the checkpoint-8000 subfolder within the finetuning_stage folder was selected for the next step.

4. Quantization: The fine-tuned model was quantized using W4A16 quantization. The quantization code is as provided:      
     `
      from pathlib import Path
      import torch
      from loguru import logger
      from transformers import AutoModelForCausalLM, AutoTokenizer
      
      from llmcompressor import oneshot, train
      
      # uses LLM Compressor's built-in preprocessing for ultra chat
      dataset = "ultrachat-200k"
      # Select the recipe for 2 of 4 sparsity and 4-bit activation quantization
      recipe = "2of4_w4a16_group-128_recipe.yaml"
      
      model_stub = "/mnt4/data/open_source/Qwen3-32B"
      tokenizer = AutoTokenizer.from_pretrained(model_stub)
      
      # save location of quantized model
      output_dir = "/mnt4/ckpt/xinming.wwm/Qwen3-32B-2of4_w4a16_0707_1"
      output_path = Path(output_dir)
      
      # set dataset config parameters
      splits = {"calibration": "train_gen[:10%]", "train": "train_gen"}
      max_seq_length = 256
      num_calibration_samples = 1024
      num_train_epochs = 3
      logging_steps = 500
      save_steps = 5000
      gradient_checkpointing = True  # saves memory during training
      learning_rate = 0.0001
      bf16 = True  # using full precision for training
      lr_scheduler_type = "cosine"
      warmup_ratio = 0.1
      preprocessing_num_workers = 32
      
      oneshot_kwargs = dict(
          dataset=dataset,
          recipe=recipe,
          num_calibration_samples=num_calibration_samples,
          preprocessing_num_workers=preprocessing_num_workers,
          splits=splits
      )
      
      # Oneshot quantization
      quantized_model = oneshot(
          model=(output_path / "finetuning_stage" / "checkpoint-8000"),
          **oneshot_kwargs,
          stage="quantization_stage"
      )
      
      quantized_model.save_pretrained(
          f"{output_dir}/quantization_stage_0709", skip_sparsity_compression_stats=False
      )
      tokenizer.save_pretrained(f"{output_dir}/quantization_stage_0709")
      
      logger.info(
          "llmcompressor does not currently support running ",
          "compressed models in the marlin24 format. "
          "The model produced from this example can be ",
          "run on vLLM with dtype=torch.float16.",
      )`

5. Evaluation: The quantized model was evaluated on the mmlu_pro benchmark. The evaluation results showed significantly degraded performance. mmlu_pro accuracy is 0.08.

6. 2of4_w4a16_group-128_recipe.yaml：
    sparsity_stage:
      sparsity_modifiers:
        SparseGPTModifier:
          sparsity: 0.5
          mask_structure: "2:4"
          targets: ["Linear"]
          ignore: ["re:.*lm_head"]
    finetuning_stage:
      finetuning_modifiers:
        ConstantPruningModifier:
          targets: [
            're:.*q_proj.weight',
            're:.*k_proj.weight', 
            're:.*v_proj.weight',
            're:.*o_proj.weight',
            're:.*gate_proj.weight',
            're:.*up_proj.weight',
            're:.*down_proj.weight',
          ]
          start: 0
    quantization_stage:
      quantization_modifiers:
        GPTQModifier:
          ignore: ["lm_head"]
          config_groups:
            group_0:
              weights:
                num_bits: 4
                type: "int"
                symmetric: true
                strategy: "group"
                group_size: 128
              targets: ["Linear"]
 
If we skip the fine-tuning step and directly quantize the sparsified model from the sparsity_stage folder, the evaluation results show a 10 percentage point drop compared to the original model.

**Expected behavior**
We expected the fine-tuned model to show improved accuracy, not a substantial decrease.

**Environment**
vllm==0.9.1.
CUDA Version: 12.4 A100.
Both llmcompressor and compressed_tensors are using the latest versions.

**Comment by kylesayrs** [comment]
Hi @wangwenmingaa 

> If we skip the fine-tuning step and directly quantize the sparsified model from the sparsity_stage folder, the evaluation results show a 10 percentage point drop compared to the original model.
Expected behavior
We expected the fine-tuned model to show improved accuracy, not a substantial decrease.

I'm slightly confused by your experiment description. You mention a substantial accuracy drop when evaluating the model where you "skip the fine-tuning step", but "expect the fine-tuned model to show improved accuracy". Could you please post your evaluation results of the fine tuned model?

Please note that, unlike quantization alone, sparsity often requires fine-tuning in order for the weights to readjust after sparsification.

**Comment by kylesayrs** [comment]
I'll also mention that fine tuning can require more steps than is shown in the example. See the [Axolotl integration blog post](https://developers.redhat.com/articles/2025/06/17/axolotl-meets-llm-compressor-fast-sparse-open) for more details! We'll also add a disclaimer on the example for future reference.

**Comment by wangwenmingaa** [comment]
@kylesayrs I fine-tuned the sparsified model, then quantized it after fine-tuning. The quantized model only achieved 0.08 accuracy on the MMLU_PRO dataset. I suspect issues in the fine-tuning stage, as the non-fine-tuned version performs significantly better.

**Fine-tuning Code:**
```
dataset = "ultrachat-200k"
# Select the recipe for 2 of 4 sparsity and 4-bit activation quantization
recipe = "2of4_w4a16_group-128_recipe.yaml"

# save location of quantized model
output_dir = "/mnt4/ckpt/xinming.wwm/Qwen3-32B-2of4_w4a16_0707_1"
output_path = Path(output_dir)

# set dataset config parameters
splits = {"calibration": "train_gen[:10%]", "train": "train_gen"}
max_seq_length = 512
num_calibration_samples = 512

# set training parameters for finetuning
num_train_epochs = 1
logging_steps = 500
save_steps = 2000
gradient_checkpointing = True  # saves memory during training
learning_rate = 0.0001
bf16 = True  # using full precision for training
lr_scheduler_type = "cosine"
warmup_ratio = 0.1
preprocessing_num_workers = 32
oneshot_kwargs = dict(
    dataset=dataset,
    recipe=recipe,
    num_calibration_samples=num_calibration_samples,
    preprocessing_num_workers=preprocessing_num_workers,
    splits=splits
)
training_kwargs = dict(
    bf16=bf16,
    optim="adamw_bnb_8bit",
    max_seq_length=max_seq_length,
    num_train_epochs=num_train_epochs,
    logging_steps=logging_steps,
    save_steps=save_steps,
    gradient_checkpointing=gradient_checkpointing,
    learning_rate=learning_rate,
    lr_scheduler_type=lr_scheduler_type,
    per_device_train_batch_size=4,
    warmup_ratio=warmup_ratio,
    torch_empty_cache_steps=100,
)

# Sparse finetune
train(
    model=(output_path / "sparsity_stage"),
    **oneshot_kwargs,
    **training_kwargs,
    output_dir=output_dir,
    stage="finetuning_stage",
)

```

**Comment by kylesayrs** [comment]
@wangwenmingaa I see! In your first experiment, you do sparsity + quant, SFT, then quant again. In your second experiment, you do sparsity + quant, no SFT, then quant again. Is that correct?

SFT is a difficult task for large language models, as the workflows used to trained them (and which produce the best results) are difficult to replicate. SFT with an extremely generic dataset like `ultrachat-200k` is likely to lead to loss of performance on fact-recall tasks like mmlu_pro.
